### PR TITLE
Fetch widget for `view-markdown-source` instead of maintaining a copy of it

### PR DESCRIPTION
**Pull request by yakov116**
1. LINKED ISSUES:
   https://github.com/sindresorhus/refined-github/pull/3594#issuecomment-711839565

2. TEST URLS:
   https://github.com/sindresorhus/refined-github/blob/3d517d6cd00e83a51973808c7ff03cae2b692d2b/contributing.md#L3-L5

3. SCREENSHOT:
   None

@fregante why do you say we need to exclude enterprise?

### CODE REVIEW

File: source/features/view-markdown-source.tsx [hunk 1]
```diff
@@ -119,7 +65,7 @@ async function showSource(): Promise<void> {
 	sourceButton.classList.add('selected');
 	renderedButton.classList.remove('selected');
 	blurButton(sourceButton);
-	(await source).before(lineActions());
+	(await source).before(await lineActions() as unknown as HTMLDetailsElement);
 
 	dispatchEvent(sourceButton, 'rgh:view-markdown-source');
```
Comment (fregante) [inline source/features/view-markdown-source.tsx:70]: We can dispatch this event **before** the line actions are added

File: source/features/view-markdown-source.tsx [hunk 2]
```diff
@@ -119,7 +65,7 @@ async function showSource(): Promise<void> {
 	sourceButton.classList.add('selected');
 	renderedButton.classList.remove('selected');
 	blurButton(sourceButton);
-	(await source).before(lineActions());
+	(await source).before(await lineActions() as unknown as HTMLDetailsElement);
```
Comment (fregante) [inline source/features/view-markdown-source.tsx:68]: The assertions should not be needed, just have the function return the right type

File: source/features/view-markdown-source.tsx [hunk 3]
```diff
@@ -7,73 +7,19 @@ import * as pageDetect from 'github-url-detection';
...
-					>
-						Reference in new issue
-					</a>
-				</li>
-			</ul>
-		</details-menu>
-	</details>
-));
+const lineActions = onetime(async () => {
+	// Avoid having to create the entire 70 lines of JSX. The URL is hardcoded to support enterprise
```
Comment (fregante) [inline source/features/view-markdown-source.tsx:17]: CORB will likely block this request on enterprise, but this needs to be confirmed. Either way the reason why this is hardcoded is that we just need the DOM from a file we know exists.

**Review by fregante — CHANGES_REQUESTED** [review]

File: source/features/view-markdown-source.tsx (hunk 3 above)
Comment (yakov116) [reply source/features/view-markdown-source.tsx:17]: background request?

File: source/helpers/fetch-dom.ts [hunk 4]
```diff
@@ -2,11 +2,17 @@ import mem from 'mem';
...
 	const absoluteURL = new URL(url, location.origin).toString(); // Firefox `fetch`es from the content script, so relative URLs fail
-	const response = await fetch(absoluteURL);
-	const dom = domify(await response.text());
+	let dom;
+	if (backgroundRequest) {
+		dom = domify(await browser.runtime.sendMessage({request: absoluteURL}));
+	} else {
+		const response = await fetch(absoluteURL);
+		dom = domify(await response.text());
+	}
```
Comment (yakov116) [inline source/helpers/fetch-dom.ts:14]: I don't like what I did here

File: source/background.ts [hunk 5]
```diff
@@ -45,6 +45,9 @@ addDomainPermissionToggle();
 
 // `background` fetch required to avoid avoid CORB introduced in Chrome 73 https://chromestatus.com/feature/5629709824032768
 // Don’t turn this into an `async` function https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#addListener_syntax
-browser.runtime.onMessage.addListener(({request}): Promise<Response> | void => {
-	return fetch(request).then(async response => response.json());
+browser.runtime.onMessage.addListener(({request, json}): Promise<Response> | void => {
+	return fetch(request).then(async response => {
+		const textContent = await response.text();
+		return json ? JSON.parse(textContent) : textContent;
+	});
```
Comment (fregante) [inline source/background.ts:52]: Just have this always return `.text()`. Leave JSON out

File: source/features/embed-gist-inline.tsx [hunk 6]
```diff
@@ -20,7 +20,7 @@ async function embedGist(link: HTMLAnchorElement): Promise<void> {
 
 	try {
 		// Get the gist via background.js due to CORB policies introduced in Chrome 73
-		const gistData = await browser.runtime.sendMessage({request: `${link.href}.json`});
+		const gistData = await browser.runtime.sendMessage({request: `${link.href}.json`, json: true});
 
 		const files = domify.one(gistData.div)!;
```
Comment (fregante) [inline source/features/embed-gist-inline.tsx:25]:
```suggestion
		const gistData = await browser.runtime.sendMessage({request: `${link.href}.json`});

		const files = domify.one(JSON.parse(gistData).div)!;
```

File: source/features/view-markdown-source.tsx [hunk 7]
```diff
@@ -7,73 +7,19 @@ import * as pageDetect from 'github-url-detection';
...
-						Reference in new issue
-					</a>
-				</li>
-			</ul>
-		</details-menu>
-	</details>
-));
+const lineActions = onetime(async () => {
+	// Avoid having to create the entire 60 lines of JSX. The URL is hardcoded to a file we know the DOM exists.
+	const blobToolbar = await fetchDom('https://github.com/sindresorhus/refined-github/blob/b1229bbaeb8cf071f0711bc2ed1b40dd96cd7a05/.editorconfig', '.BlobToolbar', true);
```
Comment (fregante) [inline source/features/view-markdown-source.tsx:18]: Don't taint `fetchDom` with this, it's unlikely we ever need it again.

```suggestion
	const randomKnownFile = 'https://github.com/sindresorhus/refined-github/blob/b1229bbaeb8cf071f0711bc2ed1b40dd96cd7a05/.editorconfig';
	const html = await browser.runtime.sendMessage({request: randomKnownFile});
	const blobToolbar = domify(html).querySelector('.BlobToolbar');
```

**Review by fregante — CHANGES_REQUESTED** [review]

File: source/background.ts (hunk 5 above)
Comment (fregante) [inline source/background.ts:52]: Also, I just realized that this lacked the existence check. It never returned `void`
```suggestion
browser.runtime.onMessage.addListener(({request}): Promise<Response> | void => {
	if (request) {
		return fetch(request).then(async response => response.text());
	}
```

Comment (fregante) [reply source/background.ts:52]: Also it receives an `any` which can also causes errors in the destructuring:

```suggestion
browser.runtime.onMessage.addListener((message): Promise<string> | void => {
	if (message?.request) {
		return fetch(request).then(async response => response.text());
	}
```

And the type is wrong, it's does not return a response.

**Comment by yakov116** [comment]
Working 🎉 

Thanks for confirming my thoughts
 > Don't taint `fetchDom` with this, it's unlikely we ever need it again.

File: source/background.ts [hunk 8]
```diff
@@ -45,6 +45,8 @@ addDomainPermissionToggle();
 
 // `background` fetch required to avoid avoid CORB introduced in Chrome 73 https://chromestatus.com/feature/5629709824032768
 // Don’t turn this into an `async` function https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#addListener_syntax
-browser.runtime.onMessage.addListener(({request}): Promise<Response> | void => {
-	return fetch(request).then(async response => response.json());
+browser.runtime.onMessage.addListener((message): Promise<string> | void => {
+	if (message?.request) {
+		return fetch(message.request).then(async response => response.text());
```
Comment (fregante) [inline source/background.ts:50]: Does this fail lint? I don't remember if we disabled that rule
```suggestion
		return fetch(message.request).then(response => response.text());
```

File: source/features/view-markdown-source.tsx [hunk 9]
```diff
@@ -137,7 +86,7 @@ async function showRendered(): Promise<void> {
 	sourceButton.classList.remove('selected');
 	renderedButton.classList.add('selected');
 	blurButton(renderedButton);
-	lineActions().remove();
+	(await lineActions()).remove();
 
 	dispatchEvent(sourceButton, 'rgh:view-markdown-rendered');
```
Comment (fregante) [inline source/features/view-markdown-source.tsx:91]: Same goes here
```suggestion
	dispatchEvent(sourceButton, 'rgh:view-markdown-rendered');

	(await lineActions()).remove();
```

**Review by fregante — APPROVED** [review]

File: source/background.ts (hunk 8 above)
Comment (yakov116) [reply source/background.ts:50]: No we kept is since it was only one place

**Comment by yakov116** [comment]
Is the title ok?

**Pull request merged**